1430: Extend card

seluianova opened the pull request:
### Short description
Add the ability to extend the Koblenz Pass in the app.
- The card is extendable if its expiry date is earlier than the endDate of the associated user entitlement
- The user can see a notification 3 months (90 days) before the current card expires
- When the user clicks the “Extend Card” button, he is redirected to the self-service portal

Design: https://www.figma.com/design/SDoGnXIjCCXpE1tAAJkRVc/entitlementcard?node-id=792-3416&t=kUofvHMqh07BbICq-1

### Proposed changes
- Adjust backend: add the 'extendable' field to the card verification result
- Add ExtendCardNotification widget in the app

**Not included:**
- In this PR an empty form to create card is opened (query params are not passed).
Query params will be added in (or after) https://github.com/digitalfabrik/entitlementcard/pull/1749
- Tests (separate tasks will be created)
- I have not implemented a separate view for the case when the card has already expired, as suggested in the design, because I think our current view is fine, but feel free to object.

### Side effects
No?

### Testing
TBD

### Resolved issues
Part of: #1430

seluianova commented on frontend/assets/koblenz/l10n/override_de.json:
```diff
@@ -10,6 +10,9 @@
     "codeRevoked": "Dieser Pass konnte nicht aktiviert werden, da er widerrufen wurde.",
+    "extendCard": "Pass verlängern",
+    "extendCardNotificationTitle": "Hinweis zum Ablauf des Passes",
+    "extendCardNotificationDescription": "Ihr Pass läuft demnächst ab.\nVerlängern Sie sie jetzt, um Ihre Dienste weiterhin nutzen zu können.",
```
```suggestion
    "extendCardNotificationDescription": "Ihr Pass läuft demnächst ab.\nVerlängern Sie ihn jetzt, um Ihre Dienste weiterhin nutzen zu können.",
```

seluianova commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/cards/webservice/schema/CardQueryService.kt:
```diff
@@ -34,13 +34,14 @@ class CardQueryService {
         val context = dfe.getContext<GraphQLContext>()
```
Do we need to keep the deprecated function above?

seluianova commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/cards/webservice/schema/CardQueryService.kt:
Do we need to keep the deprecated function above?

michael-markl replied:
I guess, unless it's too much overhead to maintain, we should keep it. Afaics, we can just make it a wrapper of the un-deprecated function which should minimize maintenance overhead.
If / When we have Matomo tracking, we can check if there are still devices with old app versions using it... If there are (almost) none left, we can remove it.

michael-markl commented on frontend/assets/l10n/app_de.json:
```diff
@@ -109,6 +109,9 @@
     "comparedWithID": "Ich habe die Daten mit einem amtlichen Lichtbildausweis abgeglichen.",
+    "extendCard": "Karte verlängern",
+    "extendCardNotificationTitle": "Hinweis zum Ablauf der Karte",
+    "extendCardNotificationDescription": "Ihre Karte läuft demnächst ab.\nVerlängern Sie sie jetzt, um Ihre Dienste weiterhin nutzen zu können.",
```
"Dienste" sounds a bit odd to me...

Maybe just "..., um sie weiterhin nutzen zu können." ?

seluianova replied:
what about 'um Ihre Services weiterhin nutzen zu können.' ? 
just thinking if the original meaning could be preserved

michael-markl replied:
Maybe, "um Ihre Vorteile weiterhin erhalten/nutzen zu können."? Maybe in English "... to continue receiving your benefits."?

michael-markl replied:
But maybe it only sounds odd to me and "Dienste" is also fine.

seluianova replied:
I like benefits/Vorteile

michael-markl commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/cards/webservice/schema/types/CardVerificationResultModel.kt:
```diff
@@ -4,5 +4,6 @@ import java.time.Instant
 
 data class CardVerificationResultModel(
     val valid: Boolean,
+    val extendable: Boolean = false,
```
Hmm... I think that verifiers should not be able to retrieve this information.
Possibly, we should create a new endpoint that uses the totp secret as authentication.

seluianova replied:
I assumed it's not a sensitive value but more technical, and it's not handled in the app when requested from the verifier

michael-markl replied:
I mean, I agree that for Koblenz, it does not matter because (afaik) the self-service portal would also reveal this information... Feel free to resolve it.

f1sh1918 commented:
@hauf-toni 
We should also provide a slightly adjusted message if the card was already expired, since the user may not open the app for 90 days

seluianova commented:
> We should also provide a slightly adjusted message if the card was already expired
@f1sh1918 currently it looks like this, do you think it requires adjustment?
![image](https://github.com/user-attachments/assets/b1b94c8d-d5ea-4c61-b393-ab46863dde87)

f1sh1918 commented:
> > We should also provide a slightly adjusted message if the card was already expired
> 
> @f1sh1918 currently it looks like this, do you think it requires adjustment? (there was a small discussion in the task regarding that) ![image](https://private-user-images.githubusercontent.com/115008338/385034851-b1b94c8d-d5ea-4c61-b393-ab46863dde87.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE5MjM4MDAsIm5iZiI6MTczMTkyMzUwMCwicGF0aCI6Ii8xMTUwMDgzMzgvMzg1MDM0ODUxLWIxYjk0YzhkLWQ1ZWEtNGM2MS1iMzkzLWFiNDY4NjNkZGU4Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExOFQwOTUxNDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMGMwNjg1ZjhhMjBjNmM0YzYxYmYwMDY0YmFhOTE0ZGVkZWU0ZDcwMzU0MmM2M2I1ZGQwYzRhZDcyYjNlMDAwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.XR-p4-rwrDWKmSMLJCTAN8c-3OoFx4wobmw1t4J8R9U)

Hm so if you click on "Weitere Aktionen" and click on first Item the link will differ depending on the card can be extended or not?

seluianova commented:
> Hm so if you click on "Weitere Aktionen" and click on first Item the link will differ depending on the card can be extended or not?

That was the idea, yes. Keep the existing menu item (since we have it already, and it says ‘beatragen ODER verlängern’), just add query params if the card is extendable.

seluianova commented:
When the pass expired, I adjusted the message to `Ihr Pass ist abgelaufen. Unter "Weitere Aktionen" können Sie Ihren Pass verlängern oder einen neuen Pass beantragen.` to make it suitable for when the card is expired AND extendable, and for when it is expired but NOT extendable.
I'm not sure if we need different messages (and views) for these two cases. But feel free to object.
![image](https://github.com/user-attachments/assets/067c425f-5ea5-4bba-b429-281a4ca06d06)

f1sh1918 commented:
> When the pass expired, I adjusted the message to `Ihr Pass ist abgelaufen. Unter "Weitere Aktionen" können Sie Ihren Pass verlängern oder einen neuen Pass beantragen.` to make it suitable for when the card is expired AND extendable, and for when it is expired but NOT extendable. I'm not sure if we need different messages (and views) for these two cases. But feel free to object. ![image](https://private-user-images.githubusercontent.com/115008338/387227857-067c425f-5ea5-4bba-b429-281a4ca06d06.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE5NDI5MDAsIm5iZiI6MTczMTk0MjYwMCwicGF0aCI6Ii8xMTUwMDgzMzgvMzg3MjI3ODU3LTA2N2M0MjVmLTVlYTUtNGJiYS1iNDI5LTI4MWE0Y2EwNmQwNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExOFQxNTEwMDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zYWI2MDViNjdjNDY2Njg3N2I4YzcwODU1NTM4MjgzYjU0ZTQyYWI0MDA0YzBhMGQ1ODRmOWRkODhhY2RiMzFiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.wN8hmpOA4mFQTRO0Btaexw9D6uz1MvhlnGy5gVKm0WI)

Alright thx. Gonna check your pr.
Sorry for not reading the whole pr description.
I always complain about that if other reviewers don't do that :/

f1sh1918 commented on frontend/lib/identification/card_detail_view/extend_card_notification.dart:
```diff
@@ -0,0 +1,79 @@
+                          t.identification.extendCardNotificationDescription,
+                          textScaler: TextScaler.linear(0.9),
+                        ),
+                        SizedBox(height: 8),
```
Is there a reason why you don't use the `Padding` widgets for set padding on texts and buttons?

seluianova replied:
I wondered about that, but couldn't find any reason why Padding should be prefered other than in case of personal preference.
https://stackoverflow.com/questions/50722011/sized-box-vs-padding-for-distances-in-columns-and-rows
I personally think, the SizedBox looks neater, because the Padding implies a nested structure.
It's also used in other widgets already.

seluianova replied:
I wondered about that, but couldn't find any reason why Padding should be prefered other than in case of personal preference. 
If you mean using Padding as a wrapper - then I think SizedBox would be neater because it doesn't imply a nested structure.
https://stackoverflow.com/questions/50722011/sized-box-vs-padding-for-distances-in-columns-and-rows
We also use the SizedBox in other widgets.

f1sh1918 replied:
Ok I'm fine with that even I think it's common use to add a padding around texts and is for me easier to understand the boilerplate Syntax since it's also common Syntax for every frontend dev.
Even you could also create a div Container as a spacer between components

But yes it may be personal preference

f1sh1918 commented on frontend/lib/identification/card_detail_view/extend_card_notification.dart:
```diff
@@ -0,0 +1,79 @@
+                      ],
+                    ),
+                  ),
+                  SizedBox(width: 16),
```
please use padding widget if possible

f1sh1918 commented on frontend/lib/identification/card_detail_view/extend_card_notification.dart:
```diff
@@ -0,0 +1,79 @@
+              ),
+            ),
+          )
+        : SizedBox.shrink();
```
```suggestion
        : Container();
```
I think this is the standard way to render an empty `Container()`

seluianova replied:
this is supposed to have an invisible performance benefit, but if it's confusing, I will change that

f1sh1918 commented on frontend/lib/identification/card_detail_view/extend_card_notification.dart:
```diff
@@ -0,0 +1,79 @@
+                    ),
+                  ),
+                  SizedBox(width: 16),
+                  GestureDetector(
```
There is also a standard widget from flutter that you may use (optional)
```js
                  IconButton(
                    padding: EdgeInsets.all(0),
                    alignment: Alignment.topRight,
                    onPressed: () {
                      setState(() {
                        _isVisible = false;
                      });
                    },
                    icon: const Icon(Icons.close),
                  ),
```

seluianova replied:
To make the IconButton look the same as it does now, it needs to be customized as follows:
```
                  IconButton(
                    onPressed: () {
                      setState(() {
                        _isVisible = false;
                      });
                    },
                    icon: const Icon(Icons.close),
                    iconSize: 16,
                    padding: EdgeInsets.zero,
                    constraints: const BoxConstraints(),
                    style: const ButtonStyle(
                      tapTargetSize: MaterialTapTargetSize.shrinkWrap,
                    ),
                  ),
```
While for the GestureDetector I only need to set 2 attributes.

f1sh1918 replied:
then keep it as it is

f1sh1918 reviewed (commented):
Looks really good actually.

I couldn't finish the testing, will continue tomorrow. Just left you some comments

seluianova commented:
> i think the extension container should have a box shadow too

@f1sh1918 if we add a shadow, we'll have to do something with the background color. because right now I'm using primary color + opacity.
it doesn't look nice with the shadow.

seluianova commented:
I think I have addressed all the comments, @f1sh1918 pls take a look

seluianova commented on frontend/lib/identification/card_detail_view/extend_card_notification.dart:
```diff
@@ -0,0 +1,79 @@
+                    child: Column(
+                      crossAxisAlignment: CrossAxisAlignment.start,
+                      children: [
+                        Text(t.identification.extendCardNotificationTitle),
```
applied, but I assume it depends on your PR https://github.com/digitalfabrik/entitlementcard/pull/1728
because currently the fonts don't match the design

seluianova commented:
I have removed my previous comment and added a minimal elevation to the notification to add some volume (I feel like the shadow shouldn't be the same as for the card...)
![image](https://github.com/user-attachments/assets/c4eb7295-5586-4eac-be5f-2ef2418d77ac)
I also accidentally messed up the commits a bit and did rebase, sorry if you have to reset the branch locally because of that 🙄

seluianova commented:
I removed my previous comment and added a minimal elevation to the notification to add some volume (I feel like the shadow shouldn't be the same as for the card...)
![image](https://github.com/user-attachments/assets/c4eb7295-5586-4eac-be5f-2ef2418d77ac)
I also accidentally messed up the commits a bit and did rebase, sorry if you have to reset the branch locally because of that 🙄

f1sh1918 reviewed (approved):
Works fine on ios 👍. 
Just one small thing commented

If you don't want to wait until the url is final (your other pr) then please create another ticket and add a ticket nr to the todo

seluianova commented:
> If you don't want to wait until the url is final (your other pr) then please create another ticket

I was not going to close the current task, untill everything is done. I find it convenient, but maybe we should discuss this topic in the team, if task splitting is more preferable 🙈

f1sh1918 commented:
> > If you don't want to wait until the url is final (your other pr) then please create another ticket
> 
> @f1sh1918 I was not going to close the current task, untill everything is done. I find it convenient, but maybe we should discuss this topic in the team, if task splitting is more preferable in such cases 🙈

Alright i think the best way would be to define the queryParam keys in the `index.ts(frontend)` and import it in the `config.ts(administration)`

ztefanie reviewed (commented):
I tried to test this, but I get this error:

```
lib/identification/card_detail_view/extend_card_notification.dart:92:33: Error: The argument type 'ApplicationUrl' can't be assigned to the parameter type 'String'.
 - 'ApplicationUrl' is from 'package:ehrenamtskarte/build_config/build_config.dart' ('lib/build_config/build_config.dart').
    launchUrlString(buildConfig.applicationUrl, mode: LaunchMode.externalApplication);
```

Happens both for bayern and koblenz flavor. I deleted build_config.dart file

seluianova commented:
> I tried to test this, but I get this error

@ztefanie broken after merged changed, should be fixed now

ztefanie commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/cards/webservice/schema/CardQueryService.kt:
```diff
@@ -15,32 +15,22 @@ class CardQueryService {
-        }
-        Matomo.trackVerification(context.backendConfiguration, projectConfig, context.request, dfe.field.name, cardHash, card.codeType, verificationResult)
-        return false
+        return verifyCardInProjectV2(project, card, dfe).valid
```
Do we really need to update this or can we delete this function as it is deprecated for 1.5 years and we do not use it?

seluianova replied:
There was a discussion about that above. I decided to refactor and keep it. Alternatively we should check if we don’t have users with the old app version.

ztefanie commented on frontend/assets/koblenz/l10n/override_de.json:
```diff
@@ -5,11 +5,14 @@
     "cardNotYetValid": "Der Gültigkeitszeitraum Ihres Passes hat noch nicht begonnen.",
     "checkFailed": "Ihr Pass konnte nicht auf seine Gültigkeit geprüft werden. Bitte stellen Sie sicher, dass eine Internetverbindung besteht und prüfen Sie erneut.",
     "codeRevoked": "Dieser Pass konnte nicht aktiviert werden, da er widerrufen wurde.",
+    "extendCard": "Pass verlängern",
```
It should not be needed to add the translations both to the overrides and the normal translations. As this is quite duplicate, i suggest to remove it.

seluianova replied:
They are different. Normal translation is ”Karte verlängern”, and here is “Pass verlängern”

ztefanie replied:
Ah i see, missed that
